vm_phys.c

```diff
@@ -0,0 +1,211 @@
+
+    basic_seg.start = ALIGN(basic_seg.start, PAGE_SIZE);
+    basic_seg.end = ALIGN(basic_seg.end, PAGE_SIZE);
+
+    int seg_size = (basic_seg.end - basic_seg.start) / PAGE_SIZE;
+
+    kprintf("physical memory allocator initiated: %d\n", seg_size);
+    kprintf("start of manageable memory: %p\n", basic_seg.start);
+    kprintf("end of manageable memory: %p\n", basic_seg.end);
+
+    int i;
+    for(i = 0; i < VM_NFREEORDER; i++)
+        TAILQ_INIT(&basic_seg.free_queues[i]);
+
+    /* Adding basic segments */
+    int ord = 0; /* order we're trying to find free list for */
+    int last_page = 0; /* index of last page not used */
+    for(ord = VM_NFREEORDER-1; ord >= 0; ord--)
+        while(ORDER_SIZE(ord) <= seg_size) {
+            /* Create free segment of order ord */
+            TAILQ_INSERT_TAIL(&basic_seg.free_queues[ord], &vm_page_array[last_page], freeq);
+            vm_page_array[last_page].order = ord;
+            seg_size -= ORDER_SIZE(ord);
+            last_page += ORDER_SIZE(ord);
+        }
+
+}
+
+/* Takes two pages which are buddies, and coalesce them */
+static vm_page_t coalesce(vm_page_t page1, vm_page_t page2) {
+    if(page1->order != page2->order)
+        FAIL("coalesce: trying to connect pages of different orders\n");
+
+    if(page1 < page2) {
+        if(page1 + ORDER_SIZE(page1->size) != page2)
+            FAIL("coalesce: pages are not buddies\n");
+        page1->order++;
+        return page1;
+    } else {
+        if(page2 + ORDER_SIZE(page2->size) != page1)
+            FAIL("coalesce: pages are not buddies\n");
+        page2->order++;
+        return page2;
+    }
+}
+
+static vm_page_t find_buddy(vm_page_t page) {
+    vm_page_t buddy = NULL;
+    /* Page is left buddy if it's addres is divisible
+     * by 2^(page->order+1), otherwise it's right buddy */
+
+    if( ( page - vm_page_array) % ORDER_SIZE(page->order+1) == 0 )
+        buddy = page + ORDER_SIZE(page->order);
+    else
+        buddy = page - ORDER_SIZE(page->order);
+
+    if(buddy->order != page->order)
+        return NULL;
+
+    /* Check if buddy is on freelist */
+    vm_page_t it;
+    TAILQ_FOREACH(it, &basic_seg.free_queues[page->order], freeq)
+        if(it == buddy)
+            return buddy;
+
+    return NULL;
+}
+
+/* Splits first segment of into two segments of same size */
+static int split_seg(size_t order) {
+
+    if(order >= VM_NFREEORDER || order == 0 || TAILQ_EMPTY(&basic_seg.free_queues[order]))
+        FAIL("split_seg: Cannot split segment\n");
+
+    vm_page_t page = TAILQ_FIRST(&basic_seg.free_queues[order]);
+    /* Works because page is some indice of vm_page_array */
+    vm_page_t page_buddy = &page[1 << (order-1)];
+
+    TAILQ_REMOVE(&basic_seg.free_queues[order], page, freeq);
+
+    page->order = order-1;
+    page_buddy->order = order-1;
+
+    TAILQ_INSERT_HEAD(&basic_seg.free_queues[order-1], page, freeq);
+    TAILQ_INSERT_HEAD(&basic_seg.free_queues[order-1], page_buddy, freeq);
+    return 1;
```

Comment:
If you're not planning to return anything else than 1 then don't return anything.